# Redudancy Analysis (RDA) 
### Genetic variation ~ geography + population structure

<hr style='border:2px solid gray'>

##### Adam Koller
 August 2024

### Loading R magic

In [1]:
%load_ext rpy2.ipython

### Loading libraries

In [2]:
%%R
library(vegan)
library(dplyr)

R[write to console]: Loading required package: permute

R[write to console]: Loading required package: lattice

R[write to console]: This is vegan 2.6-6.1

R[write to console]: 
Attaching package: ‘dplyr’


R[write to console]: The following objects are masked from ‘package:stats’:

    filter, lag


R[write to console]: The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




### Loading individual genotype matrix

In [3]:
%%R
snp.df = read.table('../data/genotypes/genotype_matrix.csv', sep=',', row.names=1, header=TRUE)

### Getting coordinate data

In [4]:
%%R
bams.df = read.table('../data/merged_bams/bams_4_rcr.txt', col.names=c('Sample_ID'))
bams.df$Sample_ID = sub("\\.bam$","", basename(bams.df$Sample_ID))
meta.df = read.csv('../config/meta_apriori_jittered.csv', row.names=1, header=TRUE)
coords.df = merge(bams.df, meta.df, by = "Sample_ID") %>% select(c('longitude','latitude'))

#### Transforming coords to dbMEM vectors

In [5]:
%%R
coords.dmatrix = dist(coords.df)
pcnm.out = pcnm(coords.dmatrix)
pcnm.vectors = as.data.frame(scores(pcnm.out))

### Loading ancestry coefficient data from STRUCTURE (*K* = 3)

In [6]:
%%R
q.df = read.csv('../data/structure/q_matrix_3', row.names=1)[,-1]

### Null and full baseline models for pcnm selection

In [7]:
%%R
RDA.pcnm.null = rda(snp.df ~ 1, pcnm.vectors)
RDA.pcnm.full = rda(snp.df ~ ., pcnm.vectors)

### Foward selection to select significant pcnm vectors

In [8]:
%%R
fwd.sel <- ordiR2step(RDA.pcnm.null, 
                      scope = formula(RDA.pcnm.full),
                     direction = "forward",
                     R2permutations = 1000,
                     R2scope = TRUE,
                     pstep=1000,
                     trace = 0)

In [9]:
%%R 
fwd.sel$call

Call: rda(formula = snp.df ~ PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 +
PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 +
PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + PCNM25, data =
pcnm.vectors)

              Inertia Proportion Rank
Total         310.583      1.000     
Constrained    42.233      0.136   19
Unconstrained 268.350      0.864  224
Inertia is variance 

Eigenvalues for constrained axes:
 RDA1  RDA2  RDA3  RDA4  RDA5  RDA6  RDA7  RDA8  RDA9 RDA10 RDA11 RDA12 RDA13 
8.205 6.674 2.621 2.453 2.294 2.025 1.842 1.767 1.584 1.513 1.453 1.445 1.360 
RDA14 RDA15 RDA16 RDA17 RDA18 RDA19 
1.289 1.267 1.184 1.135 1.077 1.044 

Eigenvalues for unconstrained axes:
   PC1    PC2    PC3    PC4    PC5    PC6    PC7    PC8 
3.0436 2.8710 2.7373 2.6760 2.6565 2.6369 2.5495 2.5356 
(Showing 8 of 224 unconstrained eigenvalues)



### Model with all significant PCNM vectors and q-values

In [22]:
%%R
all_variables = cbind(q.df, pcnm.vectors)

rda.full = rda(snp.df ~ PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 +
                        PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 +
                        PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + PCNM25 + prop_1 + prop_2 + prop_3,
                        data= all_variables)

In [23]:
%%R
RsquareAdj(rda.full)

$r.squared
[1] 0.1530317

$adj.r.squared
[1] 0.06871811



In [24]:
%%R
anova.cca(rda.full, step = 1000)

Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = snp.df ~ PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 + PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + PCNM25 + prop_1 + prop_2 + prop_3, data = all_variables)
          Df Variance     F Pr(>F)    
Model     22   47.529 1.815  0.001 ***
Residual 221  263.054                 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


PCNM vectors and ancestry coefficients explain 15.3% (adjusted to 6.8% for the number of explanatory variables) of the variation in individual SNP data

##### Significance of each term

In [13]:
%%R
anova.cca(rda.full, step = 1000, by='term')

Permutation test for rda under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

Model: rda(formula = snp.df ~ PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 + PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + prop_1 + prop_2 + prop_3, data = all_variables)
          Df Variance      F Pr(>F)    
PCNM1      1    5.838 4.9032  0.001 ***
PCNM4      1    5.072 4.2598  0.001 ***
PCNM3      1    4.408 3.7023  0.001 ***
PCNM5      1    2.534 2.1283  0.001 ***
PCNM9      1    2.478 2.0813  0.001 ***
PCNM2      1    2.336 1.9618  0.001 ***
PCNM6      1    2.083 1.7494  0.001 ***
PCNM7      1    1.862 1.5643  0.001 ***
PCNM8      1    1.617 1.3578  0.001 ***
PCNM10     1    1.541 1.2939  0.002 ** 
PCNM12     1    1.489 1.2510  0.003 ** 
PCNM13     1    1.487 1.2487  0.002 ** 
PCNM19     1    1.417 1.1903  0.010 ** 
PCNM34     1    1.376 1.1555  0.026 *  
PCNM29     1    1.360 1.1419  0.047 

### Partial RDA (Population structure when controlling for geography)

I also ommitted non significant variables from the full model (PCNM69, PCNM20, and prop_3)

In [25]:
%%R
p.RDA.struct = rda(snp.df ~ prop_1 + prop_2 + 
                            Condition(PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 
                                      + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + 
                                      PCNM13 + PCNM19 + PCNM34 + PCNM29 + PCNM15 + 
                                      PCNM69 + PCNM20 + PCNM25), data = all_variables)

In [26]:
%%R
RsquareAdj(p.RDA.struct)

$r.squared
[1] 0.01292869

$adj.r.squared
[1] 0.00570746



##### Significance of each term

In [27]:
%%R
anova.cca(p.RDA.struct, step = 1000, by='term')

Permutation test for rda under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

Model: rda(formula = snp.df ~ prop_1 + prop_2 + Condition(PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 + PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + PCNM25), data = all_variables)
          Df Variance      F Pr(>F)    
prop_1     1    1.958 1.6443  0.001 ***
prop_2     1    2.058 1.7280  0.001 ***
Residual 222  264.334                  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


##### Significance of entire model

In [28]:
%%R
anova.cca(p.RDA.struct, step = 1000)

Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = snp.df ~ prop_1 + prop_2 + Condition(PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 + PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + PCNM25), data = all_variables)
          Df Variance      F Pr(>F)    
Model      2    4.015 1.6862  0.001 ***
Residual 222  264.334                  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1


### Partial RDA (Geography when controlling for population structure)

In [29]:
%%R
p.RDA.geo = rda(snp.df ~ PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 + PCNM34 + 
                PCNM29 + PCNM15 + PCNM69 + PCNM20 + 
                PCNM25 + 
                            Condition(prop_1 + prop_2), data = all_variables)

In [30]:
%%R
RsquareAdj(p.RDA.geo)

$r.squared
[1] 0.09760056

$adj.r.squared
[1] 0.02496484



##### Significance of each term

In [32]:
%%R
anova.cca(p.RDA.geo, step = 1000, by='term')

Permutation test for rda under reduced model
Terms added sequentially (first to last)
Permutation: free
Number of permutations: 999

Model: rda(formula = snp.df ~ PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 + PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + PCNM25 + Condition(prop_1 + prop_2), data = all_variables)
          Df Variance      F Pr(>F)    
PCNM1      1    2.295 1.9274  0.001 ***
PCNM4      1    1.974 1.6582  0.001 ***
PCNM3      1    1.570 1.3187  0.001 ***
PCNM5      1    1.755 1.4738  0.001 ***
PCNM9      1    2.138 1.7955  0.001 ***
PCNM2      1    2.009 1.6871  0.001 ***
PCNM6      1    1.837 1.5431  0.001 ***
PCNM7      1    1.631 1.3695  0.001 ***
PCNM8      1    1.462 1.2275  0.001 ***
PCNM10     1    1.444 1.2127  0.001 ***
PCNM12     1    1.402 1.1776  0.002 ** 
PCNM13     1    1.480 1.2432  0.001 ***
PCNM19     1    1.379 1.1582  0.005 ** 
PCNM34     1    1.361 1.1431  0.006 ** 
PCNM29     1    1.338 1.1

##### Significance of entire model

In [31]:
%%R
anova.cca(p.RDA.geo, step = 1000)

Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = snp.df ~ PCNM1 + PCNM4 + PCNM3 + PCNM5 + PCNM9 + PCNM2 + PCNM6 + PCNM7 + PCNM8 + PCNM10 + PCNM12 + PCNM13 + PCNM19 + PCNM34 + PCNM29 + PCNM15 + PCNM69 + PCNM20 + PCNM25 + Condition(prop_1 + prop_2), data = all_variables)
          Df Variance      F Pr(>F)    
Model     19   30.313 1.3399  0.001 ***
Residual 222  264.334                  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
